### ISSUE TITLE
Is there a way to speed up the initial rebalance with multiple threads and BaseConsumer?

### ISSUE THREAD

**Issue by jcaesar** [fede1024/rust-rdkafka#145]
I have a situation where I need to know from when on messages to a topic will be picked up by my consumer, so I can trigger another process to start writing to that topic after that, ensuring I will receive all its messages.

Currently, I create a BaseConsumer in my main thread, and then start polling that BaseConsumer from a bunch of worker threads. I install a ConsumerContext to listen for a rebalance, and once I receive the first rebalance, I assume that all further messages will be received by my consumers. That seems to work, but takes 3-10 seconds on start-up (depending on the weather), which I find a bit too long.

I've tried to speed things up by using `assign` immediately after creating the consumer and subscribing like so:
```rust
let m = consumer.fetch_metadata(Some(&topic), Duration::from_secs(3)).ok()?;
let t = m.topics().iter().find(|t| t.name() == topic)?;
let tpl: Option<HashMap<(String, i32), Offset>> = t.partitions().iter().map(|p|
    Some(((topic.to_owned(), p.id()), Offset::Offset(
        consumer.fetch_watermarks(&topic, p.id(), Duration::from_millis(200)).map(|(_, high)| high).ok()?
    )))
).collect();
consumer.assign(&tpl).ok()?;
Some(())
```
and then I don't receive any messages at all (it seems).

Can I somehow trigger a rebalanced or force consumption from a certain set of offsets?

(Sorry for the blatant support request. :/ If it's a "not implemented", I shall try…)

**Comment by gklijs** [comment]
From librdkafka https://docs.confluent.io/3.0.0/clients/librdkafka/classRdKafka_1_1KafkaConsumer.html#af3f872bcb7d46febbaea63518f854b46 exposed as https://docs.rs/rdkafka/0.21.0/rdkafka/consumer/trait.Consumer.html#method.assign. Didn't try it myself but should work.

**Comment by jcaesar** [comment]
I'm guessing it makes a difference from which thread assign is being called? Because I'm using assign (as seen above), but that just gets me a rebalance with everything unassigned.

Guess I'll try calling it from each consumer thread with only a subset of partitions.

**Comment by gklijs** [comment]
Could be you should **not** set a consumer group id for it to work, since it will rebalance when it does.

**Comment by jcaesar** [comment]
I am not setting a consumer group id.

**Comment by jcaesar** [comment]
So I'm not
- not setting a consumer group id
- retrieving the partition offsets
- either calling `subscribe` or not
- splitting up the offset list and passing it to my threads, along with an `Arc<BaseConsumer>`
- calling `assign` from the threads

Depending on whether I subscribe, I either get a [`Rebalance::Revoke`](https://docs.rs/rdkafka/0.21.0/rdkafka/consumer/enum.Rebalance.html#variant.Revoke) or simply no messages at all.

**Comment by benesch** [comment]
There's a configuration parameter that controls the initial group rebalance delay: `group.initial.rebalance.delay.ms`. You can read more about it here: https://cwiki.apache.org/confluence/display/KAFKA/KIP-134%3A+Delay+initial+consumer+group+rebalance.

If you assign topics manually, then my understanding is you won't be getting a rebalance callback. I believe that's expected behavior. You definitely don't want to be calling assign from multiple threads, because the assignments will just overwrite each other.

I'm afraid there's not much more I can say without a full reproduction.

**Comment by benesch** [comment]
This turned out to be an upstream bug. See edenhill/librdkafka#2735 and edenhill/librdkafka#2865. Let me know if you're still having trouble with the latest version of rust-rdkafka!

**Comment by jcaesar** [comment]
Hm.

So indeed v0.24 takes 3.1 seconds where v0.23 took 6.0 seconds. But looking at some of the commits you linked transitively, they don't seem to be included in librdkafka v1.4. So I tried rust-rdkafka and librdkafka on master (d772e0e and e116a873). That took it back to 6 seconds.

It's possible that I did something wrong when mucking around with the linker options. Either way, if this does regress again, I'll build a reproducer for librdkafka and file an issue there. Thank you.
